java-util/src/main/java/io/druid/java/util/common/io/NativeIO.java

```diff
@@ -0,0 +1,168 @@
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.java.util.common.io;
+
+import java.io.File;
+import java.io.FileDescriptor;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.RandomAccessFile;
+import java.lang.reflect.Field;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.sun.jna.LastErrorException;
+import com.sun.jna.Native;
+import com.sun.jna.Platform;
+
+/**
+ * Native I/O operations in order to minimize cache impact.
+ */
+public class NativeIO
+{
+  private static final Logger LOG = LoggerFactory.getLogger(NativeIO.class);
+
+  private static final int POSIX_FADV_DONTNEED = 4; /* fadvise.h */
+
+  private static boolean initialized = false;
+  private static boolean fadvisePossible = true;
+
+  static {
+    try {
+      Native.register(Platform.C_LIBRARY_NAME);
+      initialized = true;
+    }
+    catch (NoClassDefFoundError e) {
+      LOG.info("JNA not found. Native methods will be disabled.");
+    }
+    catch (UnsatisfiedLinkError e) {
+      LOG.info("Unable to link C library. Native methods will be disabled.");
+    }
+    catch (NoSuchMethodError e) {
+      LOG.warn("Obsolete version of JNA present; unable to register C library");
+    }
+  }
+
+  public static native int posix_fadvise(int fd, long offset, long len, int flag) throws LastErrorException;
+
+  private NativeIO() {}
+
+  private static Field getFieldByReflection(Class cls, String fieldName)
+  {
+    Field field = null;
+
+    try {
+      field = cls.getDeclaredField(fieldName);
+      field.setAccessible(true);
+    }
+    catch (Exception e) {
+      // We don't really expect this so throw an assertion to
+      // catch this during development
+      assert false;
+      LOG.warn("Unable to read {} field from {}", fieldName, cls.getName());
+    }
+
+    return field;
+  }
+  /**
+   * Get system file descriptor (int) from FileDescriptor object.
+   * @param descriptor - FileDescriptor object to get fd from
+   * @return file descriptor, -1 or error
+   */
+  public static int getfd(FileDescriptor descriptor)
+  {
+    Field field = getFieldByReflection(descriptor.getClass(), "fd");
+    try {
+      return field.getInt(descriptor);
+    }
+    catch (Exception e) {
+      LOG.warn("Unable to read fd field from java.io.FileDescriptor");
+    }
+
+    return -1;
+  }
+
+  /**
+   * Remove pages from the file system page cache when they wont
+   * be accessed again
+   *
+   * @param fd     The file descriptor of the source file.
+   * @param offset The offset within the file.
+   * @param len    The length to be flushed.
+   */
+
+  public static void trySkipCache(int fd, long offset, long len)
+  {
+    if (!initialized || !fadvisePossible || fd < 0) {
+      return;
+    }
+    try {
+      posix_fadvise(fd, offset, len, POSIX_FADV_DONTNEED);
```

Comment:
can you put a comment here on why the return value is ignored?

Reply:
Sure.